---
title: "PloidySelection_Figs"
author: "Patrick Monnahan"
date: "10/25/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
library(assertthat)
library(tidyverse)
library(PopGenome)
library(stringr)
library(data.table)
library(gridExtra)

```

HW Genotype frequencies for diploids and tetraploids
```{r}
# Generate HW genotype frequencies for dips and tets
genoFreqs = function(freqs){
  AA = freqs ^ 2
  Aa = 2 * freqs * (1 - freqs)
  aa = (1 - freqs) ^ 2
  AAAA = freqs ^ 4
  AAAa = 4 * (freqs ^ 3) * (1 - freqs)
  AAaa = 6 * (freqs ^ 2) * (1 - freqs) ^ 2
  Aaaa = 4 * freqs * (1 - freqs) ^ 3
  aaaa = (1 - freqs) ^ 4
  dfreq = data.frame("p" = freqs, "AA" = AA, "Aa" = Aa, "aa" = aa, "Ploidy" = rep(2, length(freqs)))
  tfreq = data.frame("p" = freqs, "AAAA" = AAAA, "AAAa" = AAAa, "AAaa" = AAaa, "Aaaa" = Aaaa, "aaaa" = aaaa, "Ploidy" = rep(4, length(freqs)))
  DF = rbind(melt(dfreq, id.var = c("p", "Ploidy")), melt(tfreq, id.var = c("p", "Ploidy")))
  return(DF)
}

# Calculate diploid/tetraploid genotype frequencies across entire range of allele frequency values
freqs = genoFreqs(seq(0,1,0.01))

# Plot HW genotype frequencies
ggplot(freqs, aes(x = p, y = value, color = as.factor(Ploidy), group = variable)) + geom_line() + scale_color_manual(name="Ploidy", values = c('red','blue')) + annotate("text", x = 0.25, y = 0.625, label="aa", color = "red") + annotate("text", x = 0.5, y = 0.55, label="Aa", color = "red") + annotate("text", x = 0.75, y = 0.625, label="AA", color = "red") + annotate("text", x = 0.085, y = 0.6, label="aaaa", color = "blue") + annotate("text", x = 0.085, y = 0.365, label="Aaaa", color = "blue") + annotate("text", x = 0.5, y = 0.395, label="AAaa", color = "blue") + annotate("text", x = 0.915, y = 0.365, label="AAAa", color = "blue") + annotate("text", x = 0.915, y = 0.6, label="AAAA", color = "blue") + ylab("Genotype Frequency") + xlab("Allele Frequency") + theme_bw()
```

Dominance functions
```{r}
getDomCoefs = function(dominance, ploidy = 100){
  if (dominance == 0){
    coeffs = rep(0, ploidy - 1)
  }
  else if (dominance == 1){
    coeffs = rep(1, ploidy - 1)
  }
  else {
    dom = (1 - dominance) - 0.5
    dom = abs((dom * 10 + sign(dom)) ^ sign(dom))
    coeffs = (seq(1, ploidy - 1) / ploidy) ^ dom
  }
  return(coeffs)
}

doms = c(seq(0,1, 0.1))
Doms = as.data.frame(sapply(doms, getDomCoefs))
Doms = rbind(c(0,0,0,0,0,0,0), Doms, c(1,1,1,1,1,1,1))
colnames(Doms) = doms
Doms = cbind(Doms, data.frame("i/k"=seq(0,100) / 100))
mDoms = melt(Doms, id.var = "i.k")
ggplot(mDoms, aes(x = i.k, y = value, color = variable)) + geom_line() + scale_color_discrete(name = "Dominance\nScalar") + theme_bw() + xlab("i / k") + ylab("Dominance Coefficient")
```

Allele frequency trajectories plot (dips and tets)
```{r}
# Generate allele frequency trajectory based on deterministic recursion equations   
getTraj = function(s, dom, ploidy, start_freq, N = -9, end_freq = 0.99, maxGens = 9999, maxTries = 10){
  dom_coeffs = getDomCoefs(dom, ploidy)
  attempts = 0
  gen = 0
  p = start_freq
  k = ploidy
  while (attempts <= maxTries & p < end_freq & gen <= maxGens){
    df = data.frame("s" = s, "dom" = dom, "gen" = 0, "freq" = start_freq, "dp1" = 0, "w.bar" = 0, "ploidy" = ploidy)
    p = start_freq
    dp1 = 0
    gen = 0
    het_fits = 1 + dom_coeffs * s
    fits = c(1, het_fits, 1 + s) / (1 + s)
    while (p < end_freq & p > 0 & gen <= maxGens){
      q = 1 - p
      i = seq(0, k)
      Num = sum(choose(k, i) * ((k - i) / k) * (p ^ (k - i)) * (q ^ i) * rev(fits))
      Den = sum(choose(k, i) * (p ^ (k - i)) * (q ^ i) * rev(fits))
      p_prime = Num / Den
      if (N != -9){
        p_prime = sum(rbinom(N, k, p_prime)) / (k * N) 
      }
      df = rbind(df, c(s, dom, gen, p, dp1, Den, k))
      dp1 = p_prime - p
      p = p_prime
      gen = gen + 1
    }
    df = rbind(df, c(s, dom, gen, p, dp1, Den, k))
    attempts = attempts + 1
  }
  if (attempts > maxTries & nrow(df) <= maxGens){
    print(paste("Beneficial allele was lost due to drift for", maxTries, "consecutive attempts"))
  }
  if (nrow(df) > maxGens){
    print(paste("Beneficial allele did not fix before the maximum number of generations (", maxGens, ")."))
  }
  return(df[-1,])
}

#E.g
traj1 = rbind(getTraj(0.1, 0, 2, 0.05), getTraj(0.1, 0, 4, 0.05))
traj1 = rbind(traj1, getTraj(0.1, 0.5, 2, 0.05))
traj1 = rbind(traj1, getTraj(0.1, 0.5, 4, 0.05))
traj1 = rbind(traj1, getTraj(0.1, 1, 2, 0.05))
traj1 = rbind(traj1, getTraj(0.1, 1, 4, 0.05))

ggplot(traj1, aes(x=gen, y=freq, linetype=as.factor(dom), color=as.factor(ploidy))) + geom_line() + scale_color_manual(name="Ploidy",values=c('red','blue')) + scale_linetype_discrete(name="Dominance", labels=c("Recessive","Additive","Dominant")) + xlim(0,300) + theme_bw() + xlab("Generation") + ylab("Allele Frequency")

```

Allele frequency trajectories for additive allele with arbitrary ploidy
```{r}
s = 0.1
doms = c(0, 0.1, 0.3, 0.5, 0.7, 0.9, 1)
ploidies = c(2, 4, 6, 8)
traj2 = data.frame()
for (i in 1:length(doms)){
  for (j in 1:length(ploidies)){
    traj2 = rbind(traj2, getTraj(s, doms[i], ploidies[j], 0.05))
  }
}

ggplot(traj2, aes(x=gen/100, y=freq, color=as.factor(ploidy))) + geom_line() + scale_color_manual(name="Ploidy",values=c('red','blue','green','purple')) + theme_bw() + xlab("Generation (100's)") + ylab("Allele Frequency") + facet_grid(~dom, scales="free_x")
```

Mean and Variance in fitness
```{r}
# Generate fitness stats as a function of allele frequency
getFits = function(freqs, s, h1, h, h3){
  df = data.frame("s" = s, "h" = h, "freq" = s, "w.bar" = 0, "var.w" = 0, "w.bar.p" = 0, "h1" = NA, "h3" = NA, "ploidy" = 2)
  assert_that(length(h1) == length(h))
  assert_that(length(h) == length(h3))
  for (j in 1:length(s)){
    for(k in 1:length(h)){
    for (i in 1:length(freqs)){
      p = freqs[i]
      q = 1 - p
      Tfits = c(1 + s[j], 1 + (s[j] * h3[k]), 1 + (s[j] * h[k]), 1 + (s[j] * h1[k]), 1) / (1 + s[j])
      TGfreqs = c(p ^ 4, 4 * p^3 * q, 6 * p^2 * q^2, 4 * p * q^3, q ^ 4)
      Tw.bar = sum(TGfreqs * Tfits)
      Dfits = c(1 + s[j], 1 + (s[j] * h[k]), 1) / (1 + s[j])
      DGfreqs = c(p ^ 2, 2 * p * q, q ^ 2)
      Dw.bar = sum(DGfreqs * Dfits)
      Tvar.w = sum((TGfreqs * (Tfits - Tw.bar) ^ 2) / length(TGfreqs))
      Dvar.w = sum((DGfreqs * (Dfits - Dw.bar) ^ 2) / length(DGfreqs))
      Dw.bar.p = (p * Dfits[1]) + (q * Dfits[2])
      Tw.bar.p = ((p ^ 3) * Tfits[1]) + (3 * (p ^ 2) * q * Tfits[2]) + (3 * p * (q ^ 2) * Tfits[3]) + ((q ^ 3) * Tfits[4])
      df = rbind(df, c(s[j], h[k], p, Dw.bar, Dvar.w, Dw.bar.p, NA, NA, 2))
      df = rbind(df, c(s[j], h[k], p, Tw.bar, Tvar.w, Tw.bar.p, h1[k], h3[k], 4))
    }
    }
  }
  return(df[-1,])
}

fits = getFits(seq(0, 1, 0.01), c(0.1, 0.01, 0.001), c(0.25, 1, 0), c(0.5, 1, 0), c(0.75, 1, 0))
# Plot mean fitness 
mean_fit = ggplot(fits[fits$s==0.1,], aes(x=freq, y=w.bar, color = as.factor(ploidy), linetype=as.factor(h))) + geom_line() + scale_color_manual(name="Ploidy",values=c("red","blue")) + scale_linetype_discrete(name="Dominance", labels=c("Recessive","Additive","Dominant")) + xlab("Allele Frequency") + ylab("Mean Fitness") + theme_bw()

var_fit = ggplot(fits[fits$s==0.1,], aes(x=freq, y=var.w, color = as.factor(ploidy), linetype=as.factor(h))) + geom_line() + scale_color_manual(name="Ploidy",values=c("red","blue")) + scale_linetype_discrete(name="Dominance", labels=c("Recessive","Additive","Dominant")) + xlab("Allele Frequency") + ylab("Variance in Fitness") + theme_bw()

grid.arrange(mean_fit, var_fit)
```




Plot fixation probabilities as a function of ploidy
```{r}
#Fixation probs of diploids as function of N
getFix = function(p, s, N, c){
  num = 1 - exp(-2*c*N*s*p)
  den = 1 - exp(-2*c*N*s)
  return(num / den)
}

p = c(0.0001, 0.001, 0.01, 0.05, 0.1)
N = c(10, 100, 1000, 10000)
s = c(0.001, 0.01, 0.1)
c = c(2, 4, 6, 8)

params = expand.grid(p = p, s = s, N = N, ploidy = c)

params %<>% mutate(fixprob = pmap(list(p, s, N, ploidy), getFix))

params = data.frame("p" = c(0.0001, 0.001, 0.0))
fix = getFix(seq())

```

```{r}
# Function to loop over all possible combinations of s and h.
simTraj = function(s = c(0.1, 0.01, 0.001), h2 = c(0.25, 0, 1), h1 = c(0.5, 0, 1), h0 = c(0.75, 0, 1), start_freq =  0.05, end_freq = 1.0, maxGen = 9999){
  traj = data.frame()
  for (i in 1:length(s)){
    for (j in 1:length(h1)){
      print(paste("Starting; s =", s[i], ", h =", h1[j]))
      jj = dipTraj(s[i], h1[j], start_freq, end_freq, maxGen)
      kk = tetTraj(s[i], h2[j], h1[j], h0[j], start_freq, end_freq, maxGen)
      traj = rbind(traj, jj)
      traj = rbind(traj, kk)
    }
  }
  return(traj)
}

# Other plots

ggplot(traj, aes(x=gen, y=freq, linetype=as.factor(s), color=ploidy)) + geom_line()+facet_grid(~h1,scales="free_x") + scale_color_discrete()

ggplot(traj[traj$s==0.1,], aes(x=gen, y=var.w, linetype=as.factor(h1), color=ploidy)) + geom_line() + scale_color_manual(name="Ploidy",values=c("red","blue")) + scale_linetype_discrete(name="Dominance", labels=c("Recessive","Additive","Dominant")) + xlab("Generation") + ylab("Variance in Fitness") + theme_bw() + scale_x_log10()

ggplot(traj[traj$s==0.1,], aes(x=gen, y=w.bar, linetype=as.factor(h1), color=ploidy)) + geom_line() + scale_color_manual(name="Ploidy",values=c("red","blue")) + scale_linetype_discrete(name="Dominance", labels=c("Recessive","Additive","Dominant")) + xlab("Generation") + ylab("Mean Fitness") + theme_bw() + scale_x_log10()

ggplot(traj[traj$s==0.1,], aes(x=gen, y=freq, linetype=as.factor(s), color=ploidy)) + geom_line() + scale_color_manual(name="Ploidy",values=c("red","blue")) + scale_linetype_discrete(name="Dominance", labels=c("Recessive","Additive","Dominant")) + ylab("Allele Frequency") + xlab("Generation") + theme_bw() 

ggplot(fits[fits$s==0.1,], aes(x=freq, y=var.w, color = as.factor(ploidy), linetype=as.factor(h1))) + geom_line() + scale_color_manual(name="Ploidy",values=c("red","blue")) + scale_linetype_discrete(name="Dominance", labels=c("Recessive","Additive","Dominant")) + xlab("Allele Frequency") + ylab("Variance in Fitness") + theme_bw()

ggplot(fits[fits$s==0.1,], aes(x=freq, y=w.bar, color = as.factor(ploidy), linetype=as.factor(h1))) + geom_line() + scale_color_manual(name="Ploidy",values=c("red","blue")) + scale_linetype_discrete(name="Dominance", labels=c("Recessive","Additive","Dominant")) + xlab("Allele Frequency") + ylab("Mean Fitness") + theme_bw()

```

OLD or incomplete ideas
```{r}

# Attempting to implement the diffusion approximations for dips and tets...currently not working
difTraj = function(s, start_freq, end_freq, N, ploidy, max_gens = 10000){
  df = data.frame("s" = s, "gen" = 0, "freq" = start_freq, "ploidy" = ploidy)
  p = start_freq
  gen = 0
  while (p < end_freq & gen < max_gens){
    pq = p * (1 - p)
    mux = 2 * N * s * pq / tanh(2 * N * s * p)
    t = 1 / (ploidy * 2 * N)
    varx = (pq * t) ^ 0.5
    p_prime = p + mux * t
    if (runif(1) >= 0.5){
      p_prime = p_prime + varx
    }
    else{
      p_prime = p_prime - varx
    }
    df = rbind(df, c(s, gen, p_prime, ploidy))
    print(c(p_prime, gen, mux, t, mux * t, varx))
    p = p_prime
    gen = gen + 1
  }
  return(df)
}

#Functions to run mssell
traj_file = "/Users/pmonnahan/Documents/Research/PloidySim/mssel_output/mssel_input/traj_p4_s0.001_N10000_recessive_rep4-4Alt.txt"
traj_file = "/Users/pmonnahan/Documents/Research/PloidySim/testing.txt"

msselRun = function(N, n, traj_file, L = 1000000, mu = 1e-8, r = 1e-8, ploidy = 2, numWindows = 200, slideRate = 0.5, selPos = 0.5, selPop = 2, ms = "/Users/pmonnahan/Documents/Research/code/dmc/mssel_modified/mssel", ms_summarize = "/Users/pmonnahan/Documents/Research/PloidySim/scripts/ms_summarize.R"){ 
  options(scipen=999)
  p = ploidy
  
  out1 = str_replace(traj_file, ".txt", ".ms.out")
  out2 = str_replace(traj_file, ".txt", ".R.out")
  
  theta = N * p * mu * L 
  rho = N * p * r * L
  mig_str = "2 0 40 40 0 -ej 0.2825 2 1"
  
  args1 = paste((2 * p * n), 1, n * p, n * p, traj_file, L * selPos, "-r", rho, L, "-t", theta, "-I", mig_str)
  
  args2 = paste(out1, out2, L, numWindows, slideRate, Npops, selPop, n * p, n * p)
  print(args1)
  print(args2)
  system2(ms, args1, stdout = out1)

  system2(ms_summarize, args2)
  
  new_dat = read.table(out2)
  
  #change names
  return(new_dat)
  
}

dat = rbind(dat, new_dat)

## Stochastic simulation of selection for a beneficial allele
PloidyForSim = function(ploidy,  N, s, h, start_freq, end_freq, maxGen, maxTries){
  
  assert_that(ploidy - length(h) == 1)
  attempts = 0
  
  #Sample individuals to reproduce
  # taken from https://stats.stackexchange.com/questions/67911/how-to-sample-from-a-discrete-distribution
  sampleDist = function(PLOIDY, n, genoFreqs) { 
    sample(x = seq(0, PLOIDY), n, replace = T, prob = genoFreqs)
  }
  
  #Initalize population
  fits = sort( c(1, 1 + h * s, 1 + s) / (1 + s))
  Pop = rbinom(n = N, size = ploidy, prob = start_freq)
  p = sum(Pop) / (ploidy * N)
  freqs = c(p)
  
  while(p < end_freq & length(freqs) < maxGen & attempts < maxTries){
    #Retry if stochastic loss of beneficial allele in previous generation
    if ( p == 0 ){
      Pop = rbinom(n = N, size = ploidy, prob = start_freq)
      p = sum(Pop) / (ploidy * N)
      freqs = c(p)
      attempts = attempts + 1
    }
    
    geno_freq = c( table(factor(Pop, levels = as.character(seq(0, ploidy)))) / N )
    g_prime = ( fits * geno_freq ) / sum(fits * geno_freq) 
    # Make individuals for next generation
    p1 = sampleDist(length(g_prime) - 1, N, g_prime)
    p2 = sampleDist(length(g_prime) - 1, N, g_prime)
    Pop = c()
    for (i in 1:N){
      g1 = sample(c(rep(0, ploidy - p1[i]), rep(1, p1[i])), ploidy / 2, replace = F)
      g2 = sample(c(rep(0, ploidy - p2[i]), rep(1, p2[i])), ploidy / 2, replace = F)
      Pop = c(Pop, sum(g1) + sum(g2))
    }
    p = sum(Pop) / (ploidy * N)
    print(p)
    freqs = c(freqs, p)
  }
  return(freqs)
}

writeTraj = function(file, traj, numPops, selPops, timeScale, trajName = "rep1"){
  gen = seq(1, length(traj) + 1) / timeScale
  traj = rev(traj)
  traj = c(traj, 0)
  anc_traj = rep(0, length(traj))
  fileConn = file(file)
  # writeLines(c(paste("ntraj:", "1"), paste("npop:", numPops), "", paste("n:", length(traj), trajName), paste(gen, traj, anc_traj), sep="\t"), fileConn)
  writeLines(c(paste("ntraj:", "1"), paste("npop:", numPops), paste("n:", length(traj), trajName)), con = fileConn)
  
  dat = data.frame("gen"=gen, "traj"=traj, "anc"=anc_traj)
  print(dat)
  write.table(dat, file, quote=F, col.names = F, row.names = F, sep="\t", append=T)
  close(fileConn)
}
```


